### PR TITLE
feat(plugins): Typing Indicator

### PR DESCRIPTION
**Pull request by Nuckyz**

### CODE REVIEW

File: src/plugins/typingIndicator.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import definePlugin, { OptionType } from "@utils/types";
+import { find, findStoreLazy } from "@webpack";
+import { ChannelStore, GuildMemberStore, Tooltip, UserStore, useStateFromStores } from "@webpack/common";
+
+const ThreeDots = LazyComponent(() => find(m => m.type?.render?.toString()?.includes("().dots")));
+const TypingStore = findStoreLazy("TypingStore");
+const UserGuildSettingsStore = findStoreLazy("UserGuildSettingsStore");
+
+function getDisplayName(guildId: string, userId: string) {
+    return GuildMemberStore.getMember(guildId, userId)?.nick ?? UserStore.getUser(userId).username;
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:32]:
```suggestion
    return GuildMemberStore.getNick(guildId, userId) ?? UserStore.getUser(userId).username;
```

also how about friend nicknames?

File: src/plugins/typingIndicator.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+function TypingIndicator({ channelId }: { channelId: string; }) {
+    const guildId = ChannelStore.getChannel(channelId).guild_id;
+
+    if (!settings.store.includeMutedChannels) {
+        const isChannelMuted = UserGuildSettingsStore.isChannelMuted(guildId, channelId);
+        if (isChannelMuted) return null;
+    }
+
+    const typingUsers: Record<string, number> = useStateFromStores(
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:43]: using hooks after a potential early return is dangerous, you should probably move this above the early return

File: src/plugins/typingIndicator.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        [TypingStore],
+        () => ({ ...TypingStore.getTypingUsers(channelId) as Record<string, number> }),
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:45]: why do you spread here?

File: src/plugins/typingIndicator.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        case 3: {
+            tooltipText = `${getDisplayName(guildId, typingUsersArray[0])}, ${getDisplayName(guildId, typingUsersArray[1])} and ${getDisplayName(guildId, typingUsersArray[2])} are typing...`;
+            break;
+        }
+        default: {
+            tooltipText = `${getDisplayName(guildId, typingUsersArray[0])}, ${getDisplayName(guildId, typingUsersArray[1])}, ${getDisplayName(guildId, typingUsersArray[2])} and ${typingUsersArray.length - 3} other${(typingUsersArray.length - 3) !== 1 ? "s" : ""} are typing...`;
+            break;
+        }
+    }
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:76]: it would be much nicer to use Discord's strings for this instead, that way it's localised

**Review by Vendicated — CHANGES_REQUESTED** [review]

File: src/plugins/typingIndicator.tsx (hunk 3 above)
Comment (Nuckyz) [reply src/plugins/typingIndicator.tsx:45]: I want to make sure discord doesn't modify the object before I do the checks

File: src/plugins/typingIndicator.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        );
+    }
+
+    return null;
+}
+
+const settings = definePluginSettings({
+    includeMutedChannels: {
+        type: OptionType.BOOLEAN,
+        description: "Wheter to show the typing indicator for muted channels.",
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:109]:
```suggestion
        description: "Whether to show the typing indicator for muted channels.",
```

File: src/plugins/typingIndicator.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+*/
+
+import { definePluginSettings } from "@api/settings";
+import ErrorBoundary from "@components/ErrorBoundary";
+import { Devs } from "@utils/constants";
+import { LazyComponent } from "@utils/misc";
+import definePlugin, { OptionType } from "@utils/types";
+import { find, findLazy, findStoreLazy } from "@webpack";
+import { ChannelStore, GuildMemberStore, Tooltip, UserStore, useStateFromStores } from "@webpack/common";
+import { Settings } from "Vencord";
```
Comment (Vendicated) [inline src/plugins/typingIndicator.tsx:26]: 🤨

**Review by Vendicated — APPROVED** [review]

**Pull request merged**